// @melt-ui/svelte/dist/builders/dropdown-menu/create.js
import { overridable, toWritableStores } from '../../internal/helpers/index.js';
import { writable } from 'svelte/store';
import { createMenuBuilder } from '../menu/index.js';
import { omit } from '../../internal/helpers/object';
const defaults = {
    arrowSize: 8,
    positioning: {
        placement: 'bottom',
    },
    preventScroll: true,
    closeOnEscape: true,
    closeOnOutsideClick: true,
    portal: undefined,
    loop: false,
    dir: 'ltr',
    defaultOpen: false,
    forceVisible: false,
    typeahead: true,
    closeFocus: undefined,
    disableFocusFirstItem: false,
    closeOnItemClick: true,
};
export function createDropdownMenu(props) {
    const withDefaults = { ...defaults, ...props };
    const rootOptions = toWritableStores(omit(withDefaults, 'ids'));
    const openWritable = withDefaults.open ?? writable(withDefaults.defaultOpen);
    const rootOpen = overridable(openWritable, withDefaults?.onOpenChange);
    const rootActiveTrigger = writable(null);
    const nextFocusable = writable(null);
    const prevFocusable = writable(null);
    const { trigger, menu, item, arrow, createSubmenu, createCheckboxItem, createMenuRadioGroup, separator, group, groupLabel, ids, } = createMenuBuilder({
        rootOptions,
        rootOpen,
        rootActiveTrigger,
        nextFocusable,
        prevFocusable,
        selector: 'dropdown-menu',
        removeScroll: true,
        ids: withDefaults.ids,
    });
    return {
        ids,
        elements: {
            trigger,
            menu,
            item,
            arrow,
            separator,
            group,
            groupLabel,
        },
        states: {
            open: rootOpen,
        },
        builders: {
            createCheckboxItem,
            createSubmenu,
            createMenuRadioGroup,
        },
        options: rootOptions,
    };
}
